File: 54-spiral-matrix/54-spiral-matrix.cpp
class Solution {
public:
    

    vector<int> spiralOrder(vector<vector<int>>& arr) {
        int n  = arr.size(),m= arr[0].size();
        vector<int> ans;
        int row_start=0,row_end=n-1,col_start=0,col_end=m-1;
    while(row_start<=row_end and col_start<=col_end){
        for(int i = col_start;i<=col_end;i++){
            ans.push_back(arr[row_start][i]);
        }
       
        row_start++;
        
           if(row_start>row_end)break;
            for(int i = row_start;i<=row_end;i++){
            ans.push_back(arr[i][col_end]);
            }
        
       
        col_end--;
        
        if(col_start>col_end)break;
            for(int i =col_end ; i>=col_start ; i--){
            ans.push_back(arr[row_end][i]);
            }
        
       
        row_end--;
         
        for(int i= row_end ; i>=row_start ;i--){
            ans.push_back(arr[i][col_start]);
        }
        
        col_start++;
        
    }
        return ans;
    }
};

File: 54-spiral-matrix/54-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] nums) {
        List<Integer> ans = new ArrayList<Integer>();
        int n = nums.length;
        int m = nums[0].length;
        int rstart =0 , rend = n-1;
        int cstart = 0 , cend = m-1;
        int dir =0;
        while(rstart<=rend && cstart<=cend){
            if(dir==0){
                for(int i=cstart;i<=cend;i++){
                    ans.add(nums[rstart][i]);
                }
                rstart++;
            }
            else if(dir==1){
                for(int i=rstart;i<=rend;i++){
                    ans.add(nums[i][cend]);
                }
                cend--;
            }
            else if(dir==2){
                for(int i=cend;i>=cstart;i--){
                    ans.add(nums[rend][i]);
                }
                rend--;
            }
            else if(dir==3){
                for(int i=rend;i>=rstart;i--){
                    ans.add(nums[i][cstart]);
                }
                cstart++;
            }
            
            dir = (dir+1)%4;
        }
        return ans;
    }
}